File: backend/src/jobs/scanScheduler.ts
import cron from 'node-cron';
import type { ScheduledTask } from 'node-cron';
import { ScanTaskService } from '../services/ScanTaskService.js';
import { executeScanTask } from '../services/ScanTaskExecutor.js';
import { logger } from '../config/logger.js';

/**
 * 保存当前运行的定时任务实例（支持多个任务）
 * 用于管理和控制定时任务的启动、停止和重启
 */
let currentSchedulers: ScheduledTask[] = [];

/**
 * 执行数据库中的定时任务
 * 根据任务ID执行对应的扫描任务
 */
async function executeDatabaseTask(taskId: number): Promise<void> {
  try {
    const taskService = new ScanTaskService();
    const task = await taskService.getTaskById(taskId);

    if (!task) {
      logger.error(`[定时任务] 任务 ID ${taskId} 不存在`);
      return;
    }

    if (!task.enabled) {
      logger.debug(`[定时任务] 任务 ${task.name} 已禁用，跳过执行`);
      return;
    }

    logger.info(`[定时任务] 开始执行数据库任务: ${task.name}`);

    // 执行任务
    const result = await executeScanTask(task, { taskId: task.id });

    // 更新最后执行时间
    await taskService.updateLastExecuteTime(task.id, Date.now());

    if (result.success) {
      logger.info(`[定时任务] 任务 ${task.name} 执行成功，扫描 ${result.scannedRepositories.length} 个仓库，新增 ${result.totalCommits} 条提交`);
    } else {
      logger.error(`[定时任务] 任务 ${task.name} 执行失败: ${result.errorMessage}`);
    }
  } catch (error) {
    const errorMsg = error instanceof Error ? error.message : String(error);
    logger.error(`[定时任务] 执行数据库任务失败: ${errorMsg}`);
  }
}

/**
 * 启动定时任务调度器
 * 
 * 功能说明：
 * 1. 如果已有定时任务在运行，先停止旧任务
 * 2. 从数据库加载启用的定时任务
 * 3. 支持多个定时任务同时运行
 *
 * 数据库任务：
 * - 从 ScanTask 表中加载 taskType='scheduled' 且 enabled=true 的任务
 * - 每个任务使用自己的 Cron 表达式和扫描配置
 *
 * 注意：
 * - 启动时只注册调度器，不执行额外扫描
 * - 日常扫描依赖用户手动触发或定时任务按计划执行
 */
export async function startScheduler(): Promise<void> {
  // 如果已有定时任务在运行，先停止（避免重复启动）
  if (currentSchedulers.length > 0) {
    logger.info('[定时任务] 检测到已有任务运行，先停止旧任务');
    stopScheduler();
  }

  const taskService = new ScanTaskService();
  const schedulers: ScheduledTask[] = [];

  // 加载数据库中的定时任务
  try {
    const dbTasks = await taskService.getEnabledScheduledTasks();

    if (dbTasks.length > 0) {
      dbTasks.forEach((task) => {
        if (!task.cronExpression) {
          logger.warn(`[定时任务] 任务 ${task.name} 没有 Cron 表达式，跳过`);
          return;
        }

        const taskHandler = () => {
          executeDatabaseTask(task.id).catch((error) => {
            logger.error(`[定时任务] 执行数据库任务 ${task.name} 失败:`, error);
          });
        };

        const scheduledTask = cron.schedule(task.cronExpression, taskHandler, {
          timezone: 'Asia/Shanghai'  // 显式设置时区，确保定时任务按北京时间执行
        });

        schedulers.push(scheduledTask);
        logger.info(
          `[定时任务] 数据库任务 "${task.name}" 调度已启动，Cron 表达式: ${task.cronExpression}`
        );
      });
    }
  } catch (error) {
    logger.error('[定时任务] 加载数据库任务失败:', error);
  }

  // 检查是否有任何任务
  if (schedulers.length === 0) {
    logger.warn('[定时任务] 未找到任何定时任务，调度器未启动');
    return;
  }

  currentSchedulers = schedulers;
  logger.info(`[定时任务] 调度器已启动，共 ${currentSchedulers.length} 个定时任务`);
}

/**
 * 停止定时任务调度器
 * 
 * 功能说明：
 * 1. 停止所有正在运行的定时任务
 * 2. 清空任务列表
 * 3. 记录停止日志
 * 
 * 注意：停止后的任务无法自动恢复，需要调用 startScheduler() 重新启动
 */
export function stopScheduler(): void {
  if (currentSchedulers.length === 0) {
    logger.debug('[定时任务] 没有运行中的任务');
    return;
  }

  // 停止所有定时任务
  currentSchedulers.forEach((scheduler, index) => {
    scheduler.stop();
    logger.info(`[定时任务] 调度器 ${index + 1} 已停止`);
  });

  // 清空任务列表
  currentSchedulers = [];
  logger.info('[定时任务] 所有调度器已停止');
}

/**
 * 重启定时任务调度器
 * 
 * 功能说明：
 * 1. 停止当前所有定时任务
 * 2. 重新加载数据库任务
 * 3. 使用新配置启动定时任务
 * 
 * 使用场景：
 * - 数据库任务更新时调用
 * - 需要更新定时任务配置时调用
 * 
 * @returns {Promise<void>}
 */
export async function restartScheduler(): Promise<void> {
  logger.info('[定时任务] 正在重启调度器...');

  // 停止当前定时任务
  stopScheduler();

  // 启动新的定时任务（会读取最新数据库任务）
  await startScheduler();

  logger.info('[定时任务] 调度器已使用新配置重启');
}

File: python-backend/app/jobs/scan_scheduler.py
"""Scan task scheduler using APScheduler.

Port of backend/src/jobs/scanScheduler.ts.
Loads enabled scheduled tasks from the database and registers cron jobs.
"""

import asyncio

from apscheduler.schedulers.background import BackgroundScheduler
from apscheduler.triggers.cron import CronTrigger
from loguru import logger

from app.core.database import async_session_factory

_scheduler: BackgroundScheduler | None = None


async def _execute_database_task(task_id: int) -> None:
    """Execute a scheduled scan task by ID."""
    from app.services.scan_task_service import ScanTaskService

    task_service = ScanTaskService()

    async with async_session_factory() as session:
        try:
            task = await task_service.get_task_by_id(session, task_id)

            if task is None:
                logger.error(f"[定时任务] 任务 ID {task_id} 不存在")
                return

            if not task.get("enabled", False):
                logger.debug(f"[定时任务] 任务 {task['name']} 已禁用，跳过执行")
                return

            logger.info(f"[定时任务] 开始执行数据库任务: {task['name']}")

            from app.services.scan_task_executor import execute_scan_task

            result = await execute_scan_task(session, task, task_id=task["id"])

            import time as _time
            await task_service.update_last_execute_time(session, task["id"], int(_time.time() * 1000))

            if result.get("success"):
                repo_count = len(result.get("scannedRepositories", []))
                logger.info(
                    f"[定时任务] 任务 {task['name']} 执行成功，"
                    f"扫描 {repo_count} 个仓库，新增 {result.get('totalCommits', 0)} 条提交"
                )
            else:
                logger.error(f"[定时任务] 任务 {task['name']} 执行失败: {result.get('errorMessage')}")

            await session.commit()
        except Exception:
            await session.rollback()
            raise


def _make_task_callback(task_id: int, task_name: str, loop: asyncio.AbstractEventLoop):
    """Return a synchronous callback suitable for APScheduler."""

    def callback() -> None:
        try:
            future = asyncio.run_coroutine_threadsafe(
                _execute_database_task(task_id),
                loop,
            )
            future.result(timeout=600)
        except Exception as exc:
            logger.error(f"[定时任务] 执行数据库任务 {task_name} 失败: {exc}")

    return callback


async def start_scheduler() -> None:
    """Start the scan-task scheduler."""
    global _scheduler

    if _scheduler is not None:
        logger.info("[定时任务] 检测到已有任务运行，先停止旧任务")
        stop_scheduler()

    from app.services.scan_task_service import ScanTaskService

    task_service = ScanTaskService()
    scheduler = BackgroundScheduler(timezone="Asia/Shanghai")

    async with async_session_factory() as session:
        db_tasks = await task_service.get_enabled_scheduled_tasks(session)

    if not db_tasks:
        logger.warning("[定时任务] 未找到任何定时任务，调度器未启动")
        return

    loop = asyncio.get_running_loop()
    registered = 0

    for task in db_tasks:
        cron_expr = task.get("cronExpression")
        task_name = task.get("name", "unknown")
        if not cron_expr:
            logger.warning(f"[定时任务] 任务 {task_name} 没有 Cron 表达式，跳过")
            continue

        try:
            trigger = CronTrigger.from_crontab(cron_expr, timezone="Asia/Shanghai")
        except (ValueError, TypeError) as exc:
            logger.error(f"[定时任务] 任务 {task_name} 的 Cron 表达式 '{cron_expr}' 无效: {exc}")
            continue

        callback = _make_task_callback(task["id"], task_name, loop)
        scheduler.add_job(callback, trigger=trigger, id=f"scan_task_{task['id']}")
        registered += 1
        logger.info(f'[定时任务] 数据库任务 "{task_name}" 调度已启动，Cron 表达式: {cron_expr}')

    if registered == 0:
        logger.warning("[定时任务] 没有有效的定时任务，调度器未启动")
        return

    scheduler.start()
    _scheduler = scheduler
    logger.info(f"[定时任务] 调度器已启动，共 {registered} 个定时任务")


def stop_scheduler() -> None:
    """Stop all scheduled scan tasks and clean up."""
    global _scheduler

    if _scheduler is None:
        logger.debug("[定时任务] 没有运行中的任务")
        return

    _scheduler.shutdown(wait=False)
    _scheduler = None
    logger.info("[定时任务] 所有调度器已停止")


async def restart_scheduler() -> None:
    """Restart the scheduler with fresh configuration from the database."""
    logger.info("[定时任务] 正在重启调度器...")
    stop_scheduler()
    await start_scheduler()
    logger.info("[定时任务] 调度器已使用新配置重启")
